### ISSUE TITLE
NN product program: separate publish unit, dependency map, and acceptance program

### ISSUE THREAD

**Issue by iteathen** [iteathen/CUDA-JS#70]
## Audit reconciliation — 2026-09-01

This issue remains useful as **program/research coordination**, but its original same-package `cuda-js/nn` assumption is superseded.

Current repository authority permits an optional application-neutral NN product only as a **separate publish unit**. The published `cuda-js` core package, exports, dependencies, compatibility identity, source tree and provider-discovery behavior must remain generic. The future NN package name and source directory are intentionally unselected.

No child issue in this program is production implementation authority by itself.

## Current priority

```text
architectural disposition: accepted separate-publish-unit direction
implementation status:       NN product not implemented
qualification status:        not qualified
portfolio priority:          deferred behind CUDA-MCGS compatible-pair / current consumer-readiness path
```

## Ownership rules

- `cuda-js` core owns generic CUDA runtime/compiler/memory/operation/synchronization/provider mechanisms.
- The future NN publish unit may own training-specific tensor roles, graph semantics, autodiff, optimizer/RNG/checkpoint state, provider selection and training evidence only after accepted child specifications.
- Existing CUDA-JS-Tensor is an independent generic tensor-math library and current CUDA-MCGS evaluator dependency. Before implementing overlapping tensor/math semantics here, perform a second-instance/LEGO ownership assessment: reuse/extend CUDA-JS-Tensor when the capability is genuinely generic, or justify an independently releasable NN-specific owner. Do not silently duplicate it.
- cuBLASLt and other consumer-neutral CUDA library mechanisms remain core owners such as #75.
- NN/model/training vocabulary must not enter CUDA-JS core production contracts.

## Dependency-ready progression

When this program is explicitly reprioritized:

1. accept the separate package/source/ownership boundary;
2. reassess #72 against CUDA-JS-Tensor and define only the missing NN-specific tensor/state contract;
3. accept #73 graph semantics;
4. accept #74 autodiff/residual semantics;
5. accept memory/execution/training/provider children only in dependency order;
6. prove a complete FP32 training vertical before mixed precision, cuDNN breadth, CUDA Graph optimization or distributed scale-out;
7. native Linux and modern-hardware evidence remain exact-profile gates.

## Child issue status

- #72, #73, #74, #76-#84 are planning/research owners under this corrected boundary and are being reconciled accordingly.
- #75 is intentionally **not** an NN product child in ownership terms; it is the generic cuBLASLt provider-family extension consumed by Tensor/NN/other consumers.
- Core mechanisms already implemented elsewhere must be consumed rather than recreated under NN names.

## First vertical acceptance direction

A future first implementation remains a small complete FP32 training capsule: static model, forward, reverse-mode backward, explicit gradient accumulation, SGD, deterministic initialization/data, checkpoint/restore, independent CPU/finite-difference evidence, terminal cleanup, and no intermediate host-driven progression.

That direction is research/acceptance planning until the child authority chain is accepted.

## Non-goals

No same-package `cuda-js/nn` assumption; no duplicated generic Tensor library; no implementation from issue text alone; no Python requirement; no hidden CPU fallback; no raw provider/native handles; no distributed or mixed-precision complexity before the single-GPU correctness vertical; no impact on the current CUDA-MCGS compatible-pair critical path.

**Comment by iteathen** [comment]
## Ownership reassessment dependency — CUDA primitives below Tensor/NN

New architecture trackers **CUDA-JS#162** and **iteathen/CUDA-JS-Tensor#40** formalize a stricter layered test before this deferred NN program is implemented.

The intended stack is now explicitly assessed as:

`CUDA-JS generic GPU/runtime/provider primitives -> CUDA-JS-Tensor generic tensor semantics/planning -> optional NN semantics -> product/model consumers`.

Consequences for #70/#72-#84:

- no NN child may introduce a generic CUDA launch/scheduler/provider/compiler/resource primitive because the NN product needs it;
- no NN child may duplicate generic Tensor dtype/shape/layout/math/item-workspace semantics that naturally belong in CUDA-JS-Tensor;
- generic gaps discovered by future NN work route down to #162 or the natural Tensor owner before NN implementation proceeds;
- CUDA-JS core remains free of Tensor/NN vocabulary;
- the existence of an NN publish unit does not justify keeping a reusable CUDA mechanism in an upper library.

This also strengthens ADR-0004's **revisit trigger**. The NN issue family is already described as a separate publish unit, and CUDA-JS-Tensor now has an established independent repository/lifecycle. Before #70 authorizes production NN source, reassess whether keeping the NN publish unit in the CUDA-JS repository still has a real lifecycle/governance benefit or whether a separate repository is now the cleaner LEGO owner. That reassessment must be evidence-based; this comment does not itself move the issues or create a new repository.

#70 remains deferred. No NN implementation, package/repository move, or core API change is authorized by this update.

**Comment by iteathen** [comment]
## Repository split completed — CUDA-NN is now the semantic owner

Project-owner direction and the new `iteathen/cuda-nn` repository resolve this issue's ADR-0004 revisit trigger in favor of an independent repository/lifecycle.

The old NN program is therefore no longer active CUDA-JS work. The technical research is preserved through ownership-correct replacements rather than copied mechanically:

- program/bootstrap/authority -> `iteathen/cuda-nn#1`
- new consumer-backed inference/abstraction gate -> `iteathen/cuda-nn#2`
- #72 NN value/parameter/gradient/state roles -> `iteathen/cuda-nn#3`
- #73 NN graph -> `iteathen/cuda-nn#4`
- #74 autodiff -> `iteathen/cuda-nn#5`
- #76 NN-specific lowering -> `iteathen/cuda-nn#6`
- #77 NN memory/liveness planning -> `iteathen/cuda-nn#7`
- #78 NN data binding/path semantics -> `iteathen/cuda-nn#8`
- #79 NN execution lowering -> `iteathen/cuda-nn#9`
- #80 training state/optimizer/RNG/checkpoint semantics -> `iteathen/cuda-nn#10`
- #81 split: NN cuDNN semantic mapping -> `iteathen/cuda-nn#11`; generic provider mechanism -> CUDA-JS #163
- #82 NN fusion/provider-selection research -> `iteathen/cuda-nn#12`
- #83 qualification -> `iteathen/cuda-nn#13`, now inference-first with training separately gated
- #84 split: NN distributed-training semantics -> `iteathen/cuda-nn#14`; generic NCCL collective mechanism -> CUDA-JS #164

**#75 remains in CUDA-JS** as consumer-neutral cuBLASLt mechanism and is no longer an NN-program child in ownership terms.

Historical #71 remains closed provenance for the earlier same-repository/separate-publish-unit decision; that placement decision is superseded by the new independent repository, not erased.

Important reassessment: CUDA-NN repository creation does not itself authorize implementation. `cuda-nn#2` must first show that reusable NN inference semantics provide value beyond a product directly constructing CUDA-JS-Tensor programs. Autodiff/training is explicitly not an inference prerequisite.

CUDA-JS core remains NN/Tensor/search/product neutral. Generic lower-layer gaps route through #162 or the natural CUDA-JS owner.
